src/Client.php

```diff
@@ -68,20 +76,42 @@ public function __construct(string $endpoint, array $options = [], HttpClient $h
             ->setAllowedTypes('unfurl_media', 'bool')
             ->setAllowedTypes('allow_markdown', 'bool')
             ->setAllowedTypes('markdown_in_attachments', 'array')
+            ->setAllowedTypes('oauth_token', ['string', 'null'])
         ;
-        $this->options = $resolver->resolve($options);
+
+        $this->setOptions($options, $endpoint, $httpClient);
+    }
+
+    public function setEndpoint($endpoint, HttpClient $httpClient = null): self
+    {
+        $plugins = [
+            new BaseUriPlugin(
+                UriFactoryDiscovery::find()->createUri($endpoint)
+            ),
+        ];
+
+        if ($this->options['oauth_token']) {
+            $plugins[] = new HeaderAppendPlugin([
+                'Authorization' => 'Bearer ' . $this->options['oauth_token'],
+                'Content-Type' => 'application/json'
+            ]);
+        }
 
         $this->httpClient = new HttpMethodsClient(
             new PluginClient(
                 $httpClient ?: HttpClientDiscovery::find(),
-                [
-                    new BaseUriPlugin(
-                        UriFactoryDiscovery::find()->createUri($endpoint)
-                    ),
-                ]
+                $plugins
             ),
             MessageFactoryDiscovery::find()
         );
+        return $this;
+    }
+
+    public function setOptions(array $options, string $endpoint = self::SLACK_POST_MESSAGE_URL, HttpClient $httpClient = null): self
```

Comment:
Could you please explain why you created a `setOptions` method? I don't see the interest as the options can be defined on the constructor.

Reply:
It's a bit of a balancing act trying to not break backward-compatibility.

Slack has deprecated webhook endpoints so ideally endpoint would just be an option with a sensible default. But unfortunately, the constructor has it as the first parameter.

The two ways I see this being called are the old web-hook way:

```
$client = new Nexy\Slack\Client('https://hooks.slack.com/...', $options);
```

And the new oauth way:

```
$client = (new Nexy\Slack\Client())->setOptions($options);
```

I guess I could get rid of the setOptions function and then people would have to instantiate like this:

```
$client = new Nexy\Slack\Client(Nexy\Slack\Client::SLACK_POST_MESSAGE_URL, $options);
```

But I kinda thought that was a little yuck.